### DISCUSSION TITLE
Issue with Fault Interaction in 'Tutorial ch1_5_fault_relations.ipynb'

### DISCUSSION THREAD

**Discussion by FisPrObXu** [gempy-project/gempy#910] [Q&A] [answered]
Dear Gempy developers,

I am currently working through the Gempy tutorial ch1_5_fault_relations.ipynb. While attempting to understand the logic behind setting up fault relationships in Gempy, I noticed an issue. In my results, it seems that fault1 and fault2 do not exhibit any interaction. They appear as independent entities rather than fault2 being deformed by the influence of fault1 as shown in the official example.

I am attaching my runtime environment details and the resulting output for reference. Could you please help me understand why my faults are not interacting as expected and how to resolve this issue?

Thank you for your assistance.

Best regards,
Jingbo Xu

# APPENDIX:
1、My runtime environment
packages in environment at D:\anaconda3\envs\gempy-env:
Name                    Version                   Build  Channel      
asttokens                 2.4.1              pyhd8ed1ab_0    conda-forge
bzip2                     1.0.8                hcfcfb64_5    conda-forge
ca-certificates           2024.6.2             h56e8100_0    conda-forge
certifi                   2024.6.2                 pypi_0    pypi       
charset-normalizer        3.3.2                    pypi_0    pypi
colorama                  0.4.6              pyhd8ed1ab_0    conda-forge
comm                      0.2.2              pyhd8ed1ab_0    conda-forge
contourpy                 1.2.1                    pypi_0    pypi
cycler                    0.12.1                   pypi_0    pypi
debugpy                   1.8.1           py312h53d5487_0    conda-forge
decorator                 5.1.1              pyhd8ed1ab_0    conda-forge
exceptiongroup            1.2.0              pyhd8ed1ab_2    conda-forge
executing                 2.0.1              pyhd8ed1ab_0    conda-forge
filelock                  3.14.0                   pypi_0    pypi
fonttools                 4.53.0                   pypi_0    pypi
fsspec                    2024.6.0                 pypi_0    pypi
gempy                     2024.1.4                 pypi_0    pypi
gempy-engine              2024.1.4.2               pypi_0    pypi
gempy-plugins             2024.1.0                 pypi_0    pypi
gempy-viewer              2024.1.4                 pypi_0    pypi
idna                      3.7                      pypi_0    pypi
imageio                   2.34.1                   pypi_0    pypi
importlib-metadata        7.1.0              pyha770c72_0    conda-forge
importlib_metadata        7.1.0                hd8ed1ab_0    conda-forge
intel-openmp              2021.4.0                 pypi_0    pypi
ipykernel                 6.29.3             pyha63f2e9_0    conda-forge
ipython                   8.25.0             pyh7428d3b_0    conda-forge
jedi                      0.19.1             pyhd8ed1ab_0    conda-forge
jinja2                    3.1.4                    pypi_0    pypi
jupyter_client            8.6.2              pyhd8ed1ab_0    conda-forge
jupyter_core              5.7.2           py312h2e8e312_0    conda-forge
kiwisolver                1.4.5                    pypi_0    pypi
krb5                      1.21.2               heb0366b_0    conda-forge
lazy-loader               0.4                      pypi_0    pypi
libexpat                  2.6.2                h63175ca_0    conda-forge
libffi                    3.4.2                h8ffe710_5    conda-forge
libsodium                 1.0.18               h8d14728_1    conda-forge
libsqlite                 3.45.3               hcfcfb64_0    conda-forge
libzlib                   1.3.1                h2466b09_1    conda-forge
markupsafe                2.1.5                    pypi_0    pypi
matplotlib                3.9.0                    pypi_0    pypi
matplotlib-inline         0.1.7              pyhd8ed1ab_0    conda-forge
mkl                       2021.4.0                 pypi_0    pypi
mpmath                    1.3.0                    pypi_0    pypi
nest-asyncio              1.6.0              pyhd8ed1ab_0    conda-forge
networkx                  3.3                      pypi_0    pypi
numpy                     1.26.4                   pypi_0    pypi
openssl                   3.3.1                h2466b09_0    conda-forge
packaging                 24.0               pyhd8ed1ab_0    conda-forge
pandas                    2.2.2                    pypi_0    pypi
parso                     0.8.4              pyhd8ed1ab_0    conda-forge
pickleshare               0.7.5                   py_1003    conda-forge
pillow                    10.3.0                   pypi_0    pypi
pip                       24.0               pyhd8ed1ab_0    conda-forge
platformdirs              4.2.2              pyhd8ed1ab_0    conda-forge
plotly                    5.22.0             pyhd8ed1ab_0    conda-forge
pooch                     1.8.1                    pypi_0    pypi
prompt-toolkit            3.0.46             pyha770c72_0    conda-forge
psutil                    5.9.8           py312he70551f_0    conda-forge
pure_eval                 0.2.2              pyhd8ed1ab_0    conda-forge
pygments                  2.18.0             pyhd8ed1ab_0    conda-forge
pyparsing                 3.1.2                    pypi_0    pypi
python                    3.12.3          h2628c8c_0_cpython    conda-forge
python-dateutil           2.9.0.post0              pypi_0    pypi
python-dotenv             1.0.1                    pypi_0    pypi
python_abi                3.12                    4_cp312    conda-forge
pytz                      2024.1                   pypi_0    pypi
pyvista                   0.43.8                   pypi_0    pypi
pywin32                   306             py312h53d5487_2    conda-forge
pyzmq                     26.0.3          py312hd7027bb_0    conda-forge
requests                  2.32.3                   pypi_0    pypi
scikit-image              0.23.2                   pypi_0    pypi
scipy                     1.13.1                   pypi_0    pypi
scooby                    0.10.0                   pypi_0    pypi
setuptools                70.0.0             pyhd8ed1ab_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
stack_data                0.6.2              pyhd8ed1ab_0    conda-forge
sympy                     1.12.1                   pypi_0    pypi
tbb                       2021.12.0                pypi_0    pypi
tenacity                  8.3.0              pyhd8ed1ab_0    conda-forge
tifffile                  2024.5.22                pypi_0    pypi
tk                        8.6.13               h5226925_1    conda-forge
torch                     2.3.1                    pypi_0    pypi
torchaudio                2.3.1                    pypi_0    pypi
torchvision               0.18.1                   pypi_0    pypi
tornado                   6.4             py312he70551f_0    conda-forge
traitlets                 5.14.3             pyhd8ed1ab_0    conda-forge
typing_extensions         4.12.1             pyha770c72_0    conda-forge
tzdata                    2024.1                   pypi_0    pypi
ucrt                      10.0.22621.0         h57928b3_0    conda-forge
urllib3                   2.2.1                    pypi_0    pypi
vc                        14.3                ha32ba9b_20    conda-forge
vc14_runtime              14.38.33135         h835141b_20    conda-forge
vs2015_runtime            14.38.33135         h22015db_20    conda-forge
vtk                       9.3.0                    pypi_0    pypi
wcwidth                   0.2.13             pyhd8ed1ab_0    conda-forge
wheel                     0.43.0             pyhd8ed1ab_1    conda-forge
xz                        5.2.6                h8d14728_0    conda-forge
zeromq                    4.3.5                he1f189c_4    conda-forge
zipp                      3.17.0             pyhd8ed1ab_0    conda-forge

2、Part of resulting outputs
![comparison_of_result_1](https://github.com/gempy-project/gempy/assets/145822534/a54338cc-823d-4a82-9964-e7ed65d63225)
![comparison_of_result_2](https://github.com/gempy-project/gempy/assets/145822534/5a541e71-776d-48f8-9c1c-50d410431704)

**Comment by javoha** [comment, marked as answer]
Hi,
thanks for your question. You are definitely right - this tutorial is a little confusing. I have to go through it in more detail to see where it goes wrong and it is noted as something we need to improve. For now: Can you run the following piece of code. This should give you the desired result with a more straightforward setup. After initiating the model we define the structural frame, assign two groups as faults and set the fault relationships.

```
import gempy as gp
import gempy_viewer as gpv
from gempy_engine.core.data.stack_relation_type import StackRelationType

# Aux imports
import numpy as np
import os

#data_path = os.path.abspath('../../')

# Initiate geomodel
geo_model: gp.data.GeoModel = gp.create_geomodel(
    project_name='Faults_relations',
    extent=[0, 1000, 0, 1000, -1000, -400],
    resolution=[20, 20, 20],
    refinement=4,  # * For this model is better not to use octrees because we want to see what is happening in the scalar fields
    importer_helper=gp.data.ImporterHelper(
        path_to_orientations="tut_ch1-5_orientations.csv",
        path_to_surface_points="tut_ch1-5_points.csv",
    )
)

# Map geological series to surfaces
gp.map_stack_to_surfaces(
    gempy_model=geo_model,
    mapping_object={
        "Fault_Series1" : ('fault1'),
        "Fault_Series2": ('fault2'),
        "Strat_Series1": ('rock1', 'rock2', 'rock3', 'rock4'),
    }
)
# Defining two structural groups as faults
geo_model.structural_frame.structural_groups[0].structural_relation = StackRelationType.FAULT
geo_model.structural_frame.structural_groups[1].structural_relation = StackRelationType.FAULT

# Setting the relations
geo_model.structural_frame.fault_relations = np.array(
    [[0, 1, 1],
     [0, 0, 1],
     [0, 0, 0]]
)

# compute model
gp.compute_model(geo_model)

gpv.plot_2d(geo_model)
```
Let me know if this works for you and if you have any questions.
Cheers
Jan

**Reply by FisPrObXu** [reply]
Hi Jan,

Thank you very much for your response. After running the code you provided, I was able to observe the ideal fault relationships.

![image](https://github.com/gempy-project/gempy/assets/145822534/152ab00e-4373-4b18-97a9-184cc6008619)

I also noticed something interesting when I ran the following code in original tutorial ch1_5_fault_relations.ipynb:
```
gp.add_structural_group(
    model=geo_model,
    group_index=1,
    structural_group_name="fault_series_2",
    elements=[fault2],
    structural_relation=gp.data.StackRelationType.FAULT,
    fault_relations=gp.data.FaultsRelationSpecialCase.OFFSET_ALL
)
print(geo_model.structural_frame)

from gempy_engine.core.data.kernel_classes.solvers import Solvers

geo_model.interpolation_options.kernel_options.kernel_solver = Solvers.SCIPY_CG
geo_model.interpolation_options.kernel_options.compute_condition_number = True
gp.compute_model(gempy_model=geo_model)

```
When I set the computation backend to PYTORCH, I also obtained the ideal results.But when I use the numpy backend, the previous issues reappear. 
```
gp.compute_model(
    gempy_model=geo_model,
    engine_config=gp.data.GemPyEngineConfig(backend=gp.data.AvailableBackends.PYTORCH)
)
```

![image](https://github.com/gempy-project/gempy/assets/145822534/ed5c8bdd-f86c-4d08-a079-9fd2f197607c)

Since I'm a novice in both geology and computing, I'm not sure if this observation is useful. I hope it helps.

Thank you again for your assistance!

Best regards,

Jingbo Xu
